test/FunctionalTests/Microsoft.OData.Core.Tests/IntegrationTests/Evaluation/AutoComputePayloadMetadataInJsonIntegrationTests.cs

```diff
@@ -753,6 +752,168 @@ public void WritingDynamicComplexPropertyWithModelSpecifiedInFullMetadataMode_40
+            int deep = 0;
+            using (var messageReader = new ODataMessageReader((IODataResponseMessage)message, null, model))
+            {
+                var reader = messageReader.CreateODataResourceReader(entitySet, entityType);
+                while (reader.Read())
+                {
+                    switch (reader.State)
+                    {
+                        case ODataReaderState.ResourceStart:
+                            deep++;
+                            break;
+                        case ODataReaderState.ResourceEnd:
+                            if (deep == 2)
+                            {
+                                nestedResources.Add((ODataResource)reader.Item);
+                            }
+                            else
+                            {
+                                topLevelResource = (ODataResource)reader.Item;
+                            }
+                            deep--;
+                            break;
+                    }
+                }
+            }
+
+            Assert.NotNull(topLevelResource);
+            Assert.Equal(new Uri("http://svc/Customers(9)"), topLevelResource.Id);
+
+            Assert.Equal(2, nestedResources.Count);
+            Assert.Equal(new[] { "NS.Address", "NS.Customer" }, nestedResources.Select(n => n.TypeName));
+
+            // #1
+            ODataResource nestedResource = nestedResources.First(c => c.TypeName == "NS.Address");
+            Assert.Equal(2, nestedResource.Properties.Count());
+            Assert.Equal("Redmond", nestedResource.Properties.First(p => p.Name == "City").Value);
+
+            // #2
+            nestedResource = nestedResources.First(c => c.TypeName == "NS.Customer");
+            Assert.Equal(1, nestedResource.Properties.Count());
+            Assert.Equal(42, nestedResource.Properties.First(p => p.Name == "Id").Value);
+        }
+
+        private static IEdmModel GetModelWithAbstractType(out EdmEntitySet entitySet)
+        {
+            var model = new EdmModel();
+            var complexType = new EdmComplexType("NS", "Address");
+            complexType.AddStructuralProperty("City", EdmPrimitiveTypeKind.String);
+            complexType.AddStructuralProperty("ZipCode", EdmPrimitiveTypeKind.Int32);
+            var customer = new EdmEntityType("NS", "Customer");
+            customer.AddKeys(customer.AddStructuralProperty("Id", EdmPrimitiveTypeKind.Int32));
+            // <Property Name="ComplexData" Type="Edm.ComplexType" />
+            customer.AddStructuralProperty("ComplexData", EdmCoreModel.Instance.GetComplexType(true));
+            var container = new EdmEntityContainer("NS", "Container");
+            entitySet = container.AddEntitySet("Customers", customer);
+            model.AddElements(new IEdmSchemaElement[] { complexType, customer, container });
+
+            IEdmEntityType entityType = model.FindType("Edm.EntityType") as IEdmEntityType;

```

Comment:
>IEdmEntityType entityType = model.FindType("Edm.EntityType") as IEdmEntityType; [](start = 12, length = 79)

Why use FindType rather than EdmCoreModel.Instance.GetEntityType(true), similar to how you got the Edm.ComplexType?  Wouldn't GetEntityType be much more efficient?  I realize this is test code, but people often look at test code to figure out how to write production code, so (unless we're trying to test out a particular code path) we should use best practices in tests as well as samples, docs, etc.

Reply:
What I did here is to make sure the customer also can get the "abstract type" from the "User model".
But, I think your comments are correct. 
So, #Resolved.